Add a more friendly "latch" synchronization tool

chandlerc opened the pull request:
The standard `std::latch` is very restrictive in how it can be used, and this makes it hard to easily leverage for simple coordination between a set of dynamically scheduled tasks, where there isn't an interesting synchronizing "merge" or future result.

This tool makes it easy to establish a latch, hand out handles to it, and once all are destroyed, take whatever relevant action.

Note: this is split out of a larger change that uses it. I can wait until the use case is ready, but seemed nice to review this separately.

chandlerc committed:
Add a more friendly "latch" synchronization tool

The standard `std::latch` is very restrictive in how it can be used, and
this makes it hard to easily leverage for simple coordination between
a set of dynamically scheduled tasks, where there isn't an interesting
synchronizing "merge" or future result.

This tool makes it easy to establish a latch, hand out handles to it,
and once all are destroyed, take whatever relevant action.

danakj commented on common/latch.h:
```diff
@@ -0,0 +1,138 @@
+// A synchronization primitive similar to `std::latch` to coordinate starting
+// some action once all of a set of other actions complete.
+//
+// Users initialize the latch (with `Init), and receive a handle RAII object.
```
```suggestion
// Users initialize the latch (with `Init`), and receive a handle RAII object.
```

danakj commented on common/latch.h:
```diff
@@ -0,0 +1,138 @@
+//
+// The latch synchronizes between every destruction of a handle and the
+// destruction of the last handle, allowing code that runs after the latch is
+// satisfied to access everything written by threads that destroy a handle. For
```
```suggestion
// satisfied to access everything written by any thread that destroyed a handle. For
```

danakj commented on common/latch.h:
```diff
@@ -0,0 +1,138 @@
+  // last handle destruction. Typically, the closure here should _schedule_ the
+  // next step of work on some thread pool rather than performing it directly.
+  template <typename CallableT = decltype([] {})>
+  auto Init(CallableT on_zero = {}) -> Handle;
```
Why a template instead of receiving llvm::unique_function directly?

chandlerc replied:
Dunno, this is the way I've most often seen this written -- avoiding the implicit conversion and using a template function and the owning / erased member.

Happy to change if the other way is better?

danakj replied:
I don't have any strong feelings but Jon always pushes back against template functors, asking for the concrete type which includes the signature - though here I see we have a default value which contains a signature.

chandlerc replied:
Nah, you (and/or Jon) are right I think -- just taking `unique_function` is much cleaner. It makes it clear that it can have move-only captures, etc, and simplifies the inline vs. out-of-line code nicely. Switched. =]

danakj commented on common/latch.h:
```diff
@@ -0,0 +1,138 @@
+  Latch(const Latch&) = delete;
+  Latch(Latch&&) = delete;
+
+  // Initialize a latch and get the initial handle to it.
```
It seems it is possible to reuse a Latch and call `Init` once all previous Handles are gone. If that's intended, maybe we should document it. If it's not intended, maybe we should CHECK against it?

chandlerc replied:
I don't feel strongly here.

I don't have a use case for reuse of the latch, and so happy to disallow it. But the implementation falls out, so could go either way.

Similarly, I'm open to different names if folks would prefer in the API. Some ideas:
- `Init` (as-is)
- `Reset`
- `Obtain` (from discussing this with Hana)

Also happy to consider names other than `Handle` if the copying is too surprising with that name. Hana had suggested `Ticket` which I like, but worry a tiny bit about being confused with ticket locks and other synchronization techniques that use a ticket.

danakj replied:
I think I'd just document that it works, and add a test.

chandlerc replied:
Done.

danakj commented on common/latch_test.cpp:
```diff
@@ -0,0 +1,111 @@
+  ASSERT_FALSE(std::move(handle2).Drop());
+  EXPECT_FALSE(called);
+  ASSERT_TRUE(std::move(handle3).Drop());
+  EXPECT_TRUE(called);
```
```suggestion
  EXPECT_FALSE(std::move(handle).Drop());
  EXPECT_FALSE(called);
  EXPECT_FALSE(std::move(handle2).Drop());
  EXPECT_FALSE(called);
  EXPECT_FALSE(std::move(handle3).Drop());
  EXPECT_TRUE(called);
```
The general practice AIUI is to use ASSERT conditions to protect against crashes but otherwise allow the test to proceed as later failures can provide additional useful information.

danakj commented on common/latch_test.cpp:
```diff
@@ -0,0 +1,111 @@
+
+  // Check that dropping the new handle satisfies the latch.
+  EXPECT_FALSE(called);
+  std::move(handle2).Drop();
```
```suggestion
  EXPECT_TRUE(std::move(handle2).Drop());
```

danakj commented on common/latch_test.cpp:
```diff
@@ -0,0 +1,111 @@
+    EXPECT_FALSE(called);
+  }
+  EXPECT_FALSE(called);
+  std::move(handle).Drop();
```
```suggestion
  EXPECT_TRUE(std::move(handle).Drop());
```

danakj reviewed (approved):
LGTM

danakj commented:
Uhoh, we have no std::jthread in libc++ it looks like, so the tests will need to change a little.

chandlerc committed:
Use explicit joins rather than `jthread`.

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

chandlerc committed:
Fix

chandlerc reviewed (commented):
Doh on forgetting about the `jthread` thing.. Second time I've done that.

Should all be fixed.

chandlerc committed:
Fix more

danakj commented on common/latch_test.cpp:
```diff
@@ -87,7 +87,7 @@ TEST(LatchTest, MultiThreaded) {
   });
 
-  std::deque<std::jthread> threads;
+  std::deque<std::thread> threads;
```
```suggestion
  std::vector<std::thread> threads;
```
Is there a specific reason for deque? It just kind of says "look at me" since it's not the default choice.

chandlerc replied:
For some reason I thought I saw an error with `std::vector` and assumed it was because `std::thread` wasn't movable so switched to `std::deque`. Anyways, `std::vector` totally works fine here, done (plus the clang-tidy fix).

danakj reviewed (approved):
Still LGTM

chandlerc committed:
Document re-calling `Init` and test that.

Also switch to directly accepting an `llvm::unique_function` as that
does end up cleaner.

Also use `std::vector`, it works fine -- not sure what error made me
switch to `std::deque`.

chandlerc reviewed (commented):
Thanks -- I think all done, merging!
